### ISSUE TITLE
`GlobPatternFilterFileProvider` does not include directories of matched files

### ISSUE THREAD

**Issue by andersstorhaug** [dazinator/Dazinator.AspNet.Extensions.FileProviders#50]
This is probably best conveyed through a simple test, shown below.
The desired behavior is that, because `/subpath/foo.txt` exists, the directories `/` and `subpath` should also exist.

I'll try to find some time to look into this further, maybe I can help provide a solution :smirk:.

```csharp
[Fact]
public void Can_use_filtered_file_provider()
{
    var files = new InMemoryFileProvider();
    files.Directory.AddFile("/subpath", new StringFileInfo("bar", "foo.txt"));
    // Result: `/subpath/foo.txt`

    files.GetDirectoryContents("/").Should().NotBeEmpty();

    var filteredFiles = new GlobPatternFilterFileProvider(files, new[] {"**/*.txt"});

    filteredFiles.GetFileInfo("subpath/foo.txt").Exists.Should().BeTrue();

    // BUG
    filteredFiles.GetDirectoryContents("/").Should().NotBeEmpty();
}
```

**Comment by andersstorhaug** [comment]
It seems that passing `recurseDirectories: true` to `GlobMatchingEnumerableFileInfos` fixes the issue.

Maybe `bool recurseDirectories` can be added as a ctor param to `GlobPatternFilterFileProvider`?

**Comment by andersstorhaug** [comment]
Seems this will need more work, when wrapping a physical file provider with `recurseDirectories: true` this is very slow.
I'm guessing because there are a lot of duplicate entries but I haven't determined that yet.

**Comment by andersstorhaug** [comment]
Ok, so after working through this, for my use-case it makes sense to simply pass-on directories as-is with `GlobMatchingEnumerableFileInfos`, because it is not known whether a given directory will contain a matched file somewhere in it's subfolders.

For more context, I have an implementation that recursively enumerates all of the files in an `IFileProvider`, which is stack-based much like the implementation of the `Search` extension method here.

By not filtering out directories this still allows me to recursively enumerate a glob-filtered provider. I would imagine that the same would be true for `Search`, so it may be a change worth considering for that reason.

**Comment by dazinator** [comment]
@andersstorhaug 
Thanks for raising this!
If I am understanding this correctly - I guess there needs to be an option for whether to:

1. Include all the directories (even if they don't contain any matching files)
2. Include only directories that container results (so filter out empty directories)

Does this sound about right?

**Comment by andersstorhaug** [comment]
Yep, I think for starters (1) would allow for more use-cases (again search/recursive enumeration) so it could be a good starting point if (2) proves to be difficult.

For (2), I think that would be ideal and more likely match what folks would naturally expect, assuming sub-directories/files would be included. Though, I personally can't think of a straight-forword implementation to do it yet.

**Comment by andersstorhaug** [comment]
And off-topic, but if you were curious.

I'm using these providers to compose files together from multiple sources: multiple physical, some generated in-memory, some from embedded resources.

The `IFileProvider` abstraction seems to work pretty well for this, and providers from this library have been really useful.

**Comment by dazinator** [comment]
I finally got around to implementing this.
I just went with option 1 - and updated the readme about how the `GlobPatternFilterFileProvider` works.
So to recap:

The include / exclude glob patterns that you supply to the provider now only filter out files.
Any directory from the inner provider will be surfaced, to allow consumers to still see full directory structure, even if some of those sub directories are now empty due to files being "filtered".
